test/unit/ngsiv2/httpBindings-test.js

```diff
@@ -0,0 +1,738 @@
+                'fiware-servicepath': '/gardens'
+            }
+        },
+        groupCreation = {
+            url: 'http://localhost:4041/iot/services',
+            method: 'POST',
+            json: utils.readExampleFile('./test/groupProvisioning/provisionAliasGroup.json'),
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            }
+        };
+
+        beforeEach(function(done) {
+
+            contextBrokerMock
+            .matchHeader('fiware-service', 'smartGondor')
+            .matchHeader('fiware-servicepath', '/gardens')
+            .post('/v2/entities/Second%20UL%20Device/attrs' ,
+                utils.readExampleFile('./test/unit/ngsiv2/contextRequests/unprovisionedAliasMeasure2.json'))
+            .reply(204);
+
+            request(groupCreation, function(error, response, body) {
+                request(deviceCreation, function(error, response, body) {
+                    done();
+                    });
+                });
+            });
+
+        it('should use the device preference', function(done) {
+            request(getOptions, function(error, response, body) {
+                contextBrokerMock.done();
+                done();
+            });
+        });
+    });
+
+    describe('When a real production request arrives to the IoTA', function() {
+        var postOptions = {
+            url: 'http://localhost:' + config.http.port + '/iot/d',
+            method: 'POST',
+            qs: {
+                i: 'urn:x-iot:smartsantander:u7jcfa:fixed:t311',
+                k: '1234',
+                t: '2016-05-11T10:12:26.476659Z'
+            },
+            headers: {
+                'Content-type': 'text/plain'
+            },
+            body: 'tmp|24.4#hum|58.0#aco|0.1#apa|0.38#ao3|121.0#' +
+            'no2|115.0#pla|43.4551#plo|-3.83381#poa|28.0#spi|0.0#dia|0.0#mit|1492.0#pos|43.4630608,-3.8345434'
+        },
+        provisionProduction = {
+            url: 'http://localhost:' + config.iota.server.port + '/iot/devices',
+            method: 'POST',
+            json: utils.readExampleFile('./test/deviceProvisioning/provisionDeviceProduction.json'),
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            }
+        };
+        // Note: /v1/updateContext response is not processed by IOTA so its content is irrelevant,
+        // as far as it is a 200 OK
+        beforeEach(function(done) {
+            contextBrokerMock
+            .matchHeader('fiware-service', 'smartGondor')
+            .matchHeader('fiware-servicepath', '/gardens')
+            .post('/v1/updateContext')
+            .reply(200, '{}')
+            .post('/v2/entities/urn:x-iot:smartsantander:u7jcfa:fixed:t311/attrs')
+            .times(12)
+            .reply(204)
+            .post('/v2/entities/urn:x-iot:smartsantander:u7jcfa:fixed:t311/attrs', function(body) {
```

Comment:
It would be more clear to have an explicit .json with the expected response, as in other cases. Why we need and evaluation function in this case? Is there any dynamic parameter?

Reply:
I have done a bit of research to understand the motivation of this approach. As you can see in https://github.com/fermenreq/iotagent-ul/blob/ActiveAttsNgsi2/test/unit/ngsiv2/httpBindings-test.js#L597, in the test a group of measurement is sent to the IOTA, each one of them consists of a pair name - value, e.g., *tmp|24, hum |58*. In the test, 13 different measurements are sent.

For each one of these individual measurements, the IOTA will send an update context request to the Orion CB. For instance, for the first measurement:
```
{
    "url": "http://192.168.1.1:1026/v2/entities/urn:x-iot:smartsantander:u7jcfa:fixed:t311/attrs",
    "method": "POST",
    "headers": {
        "fiware-service": "smartGondor",
        "fiware-servicepath": "/gardens",
        "X-Auth-Token": null
    },
    "json": {
        "temperature:ambient": {
            "type": "float",
            "value": "24.4",
            "metadata": {
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2016-05-11T10:12:26.476659Z"
                }
            }
        },
        "TimeInstant": {
            "type": "DateTime",
            "value": "2016-05-11T10:12:26.476659Z"
        },
        "att_name": {
            "type": "string",
            "value": "value",
            "metadata": {
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2016-05-11T10:12:26.476659Z"
                }
            }
        },
        "coords": {
            "type": "string",
            "value": "WGS84",
            "metadata": {
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2016-05-11T10:12:26.476659Z"
                }
            }
        }
    }
}
```

This request will be different for each one of the measurements and therefore it is not possible to have a single mock for the 13 Orion CB requests. What options do we have?
1. To have 13 different mocks, each one of them with its corresponding payload. In this approach, we would validate all the requests but we would introduce some complexity in the code.
2. To have a single mock that validates just the structure of the payloads. Something like:

```
        beforeEach(function(done) {
            contextBrokerMock
            .matchHeader('fiware-service', 'smartGondor')
            .matchHeader('fiware-servicepath', '/gardens')
            .post('/v1/updateContext')            
            .reply(200, {})
            .post('/v2/entities/urn:x-iot:smartsantander:u7jcfa:fixed:t311/attrs', function(body) {
                var i = 0;
                var attributes = 0;

                for (var attribute in body) {
                    if (body.hasOwnProperty(attribute)) {
                        attributes++;
                        for (var metadata in body[attribute].metadata) {
                            if (body[attribute].metadata.hasOwnProperty(metadata)) {
                                i++;
                            }
                        }
                    }
                }
                return i === attributes - 1;
            })
            .times(13)
            .reply(204);
```

What do you think @fermenreq and @fgalan?

Reply:
Approach 2) seems to be better in this case, as there are already other tests cases with more "precise" request programmmed in the code.

A short comment whenever this kind of mocking is used (i.e. a summary or your detailed explanaiton) would be also very helpful.

Reply:
Fixed in 97b8a72905b27e3542c4abaff2b665b5487ec734